opengovsg/postmangovsg
# feat: email open tracking

## Problem

Email read receipts enhance the experience of Postman and is a frequently requested feature.

- Provides user a email read metric to let users know the reach of their campaign and the effectiveness of using email as a channel of communication
- Provides users a way of knowing if the recipient has read the email, so that they can reach the recipient through other channels if needed

Closes #1236

## Solution

**Features**:

- Setup AWS SES configuration set to receive a SNS event callback when the recipient opens the email
- Handle incoming open events and update the status of the message in the db
- Introduce `READ` status to represent emails which have been opened 

## Tests

- Deploy branch to staging, ensure workers are deployed too
- Send out an email campaign to yourself
- Ensure successful delivery is recorded in staging db for campaign messages with status `SUCCESS`
- Open the email and then check staging db for campaign messages with status `READ`
- Ensure that delivery and bounce still works as expected

## Deploy Notes
- ~Create configuration set and SNS topic and subscription for prod~
- SES configuration set (`postman-email-open`) created for both production SES region

**New environment variables**:

- `WORKER_SES_CONFIGURATION_SET` : Name of the AWS SES configuration set. Defaults to `postman-email-open`
